### PR TITLE
Added WS558, EM310, AM300 and Updated old description

### PR DESCRIPTION
**Pull request by Milesight-IoT**
#### Summary

Added WS558, EM310, AM300 and Updated old description

#### Changes

- Added .js code, -codex.yaml, .yaml and .png of WS558, EM310-UDL, AM300
- Updated firmware version numbers for several committed devices

#### Notes for Reviewers

Milesight IoT is releasing some new models of LoRa end-devices, this pull request contains the ones that haven't been added in the repository yet for users of TTI to choose. 
We noticed device models added in last pull request aren't available on TTI when adding one, kindly help on the issue and review this request to implement Milesight IoT models.

### CODE REVIEW

**Comment by rroemhild** [comment]
Is help needed? I can gladly provide some support at this PR.

**Comment by Milesight-IoT** [comment]
> Is help needed? I can gladly provide some support at this PR.

Yes, please! Your help is much appreciated.

**Comment by rroemhild** [comment]
Hi @Milesight-IoT, I opened a PR to your repo.

**Comment by Jaime-Trinidad** [comment]
@Milesight-IoT I finish the complete review, speaking with the team, is necessary to make a rollback in some changes to the firmware and hardware version, it means the ones already in the Repository must stay like this, and add the new ones, not replacing the old ones, the reason is there are users with this versions and they will change to the new firmware version when it is available for them, and making the change taking out the old one can cause problems to them. I will add each comment in the codec files because many outputs don't correspond to the bytes inputs.

**Comment by Milesight-IoT** [comment]
The problem in em300-mcs should be about door the parameter that EM300-MCS uses to check current door status, I commented the lines with door but there still seems to be problems unresolved. Kindly help and check if that's stopping validation. Thanks very much. @Jaime-Trinidad

**Comment by Milesight-IoT** [comment]
Speaking of rolling back some files, could you please tell me specifically which .yaml I should dublicate instead of replacing the old ones? @Jaime-Trinidad

**Comment by Milesight-IoT** [comment]
Hi Jaime, 

Until 4th Batch Update, there are only 2 models left from us due to unsure information, they are UC50x and UC51x. Besides there are discontinued products UC1114, UC1122, UC1152, there might be customers using them on TTN but we don't sell them or have their product URL any more. Would you suggest that I should create the files for them and leave URLs blank? Waiting for your reply, thanks.

**Comment by johanstokking** [comment]
@Milesight-IoT can you first rebase this and resolve conflicts? Also why is `vendor/index.yaml` changed? This looks like a very old version.

**Comment by Milesight-IoT** [comment]
Hi @johanstokking , I mended the conflicting files by copying latest vendor\index.yaml and collating some details in em300-mcs files. But there still seems to be conflicts, could you please check and show me what exactly they are? Many thanks.

File: vendor/index.yaml [hunk 1]
```diff
@@ -158,13 +158,14 @@ vendors:
...
     logo: decentlab_logo.svg
     ouis:
       - 70B3D57BA
     description: Decentlab is a Swiss company providing wireless sensor devices and services for distributed, cost-effective monitoring solutions. The sensor devices communicate wirelessly over LoRaWAN® and are designed for ultra low power consumption, capable of operating on batteries for several years. The devices are built for industrial applications and are ready to be deployed in any harsh indoor or outdoor environment. The service framework provides convenient access to measurement data and enables seamless integration into existing monitoring and control systems. Application areas are environmental and air quality monitoring, hydrological measurements, smart agriculture and smart cities.
     social:
-      linkedin: https://www.linkedin.com/company/decentlab/
+      linkedin: https://www.linkedin.com/company/decentlab
+      facebook: https://www.facebook.com/decentlab.gmbh
       twitter: decentlab
       github: decentlab
```
Comment (johanstokking) [inline vendor/index.yaml:170]: These are the only allowed changes here; please make sure the others are not touched

File: vendor/milesight-iot/am103&am103l.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
+
+function decodeUplink(input) {
```
Comment (johanstokking) [inline vendor/milesight-iot/am103&am103l.js:2]: Please don't use `&` in filenames

File: vendor/milesight-iot/em300-th.yaml [hunk 3]
```diff
@@ -2,14 +2,16 @@ name: EM300-TH
...
 hardwareVersions:
-  - version: 'V2.2'
+  - version: '2.x'
     numeric: 1
+  - version: '3.x'
+    numeric: 2
 
 firmwareVersions:
-  - version: '1.15'
+  - version: '1.22'
```
Comment (johanstokking) [inline vendor/milesight-iot/em300-th.yaml:11]: Existing devices provisioned with these versions will not be able to load the right codec. Is this change really necessary?

File: vendor/milesight-iot/am107.yaml [hunk 4]
```diff
@@ -1,16 +1,56 @@
...
+    numeric: 2
 
 firmwareVersions:
-  - # Firmware version
-    version: 'V1.25'
+  - version: '1.x'
     numeric: 1
     hardwareVersions:
-      - '1.4'
+      - '1.x'
```
Comment (johanstokking) [inline vendor/milesight-iot/am107.yaml:14]: Idem

File: vendor/milesight-iot/em500-co2.yaml [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
+name: Milesight EM500-CO2-LoRaWAN® Carbon Dioxide Sensor (4 in 1)
+description: EM500-CO2 is an outdoor environment monitoring sensor designed for measuring CO2, temperature, humidity and barometric pressure in harsh environments. EM500-CO2 is battery powered and designed for multiple mounting ways. It's equipped with NFC and can easily be configured by a smartphone or PC software.
 
 hardwareVersions:
   - version: 'V1.3'
     numeric: 1
 
 firmwareVersions:
-  - version: '2.28'
+  - version: '2.x'
```
Comment (johanstokking) [inline vendor/milesight-iot/em500-co2.yaml:9]: Idem

File: vendor/milesight-iot/em500-udl.yaml [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 hardwareVersions:
   - version: 'V1.3'
     numeric: 1
 
 firmwareVersions:
-  - version: '2.28'
+  - version: '2.x'
     numeric: 1
     hardwareVersions:
       - 'V1.3'
```
Comment (johanstokking) [inline vendor/milesight-iot/em500-udl.yaml:12]: Idem

File: vendor/milesight-iot/am103&am103l.js (hunk 2 above)
Comment (Milesight-IoT) [reply vendor/milesight-iot/am103&am103l.js:2]: OK, noted. May I know what's the suggested filename to use if we provide AM103 and AM103L and they share the same codec and decoder?

File: vendor/milesight-iot/em300-th.yaml (hunk 3 above)
Comment (Milesight-IoT) [reply vendor/milesight-iot/em300-th.yaml]: I rolled back the modification on the version number of firmware, and new version of hardware should remain due to updated feature with the new machine.

File: vendor/milesight-iot/am103&am103l.js (hunk 2 above)
Comment (johanstokking) [reply vendor/milesight-iot/am103&am103l.js:2]: You can do dashes or underscores

File: vendor/index.yaml [hunk 7]
```diff
@@ -158,13 +158,14 @@ vendors:
   - id: decentlab
     name: Decentlab GmbH
     vendorID: 109
-    website: https://www.decentlab.com/
+    website: https://www.decentlab.com
```
Comment (johanstokking) [inline vendor/index.yaml:161]: Can you rebase please? There shouldn't be a diff in this file other than Milesight if you make changes there.

**Comment by Milesight-IoT** [comment]
> Hi Jaime,
> 
> Until 4th Batch Update, there are only 2 models left from us due to unsure information, they are UC50x and UC51x. Besides there are discontinued products UC1114, UC1122, UC1152, there might be customers using them on TTN but we don't sell them or have their product URL any more. Would you suggest that I should create the files for them and leave URLs blank? Waiting for your reply, thanks.

Hi @johanstokking, issues about the filename should have been resolved. Could you please check out this question about Milesight discontinued products?

**Comment by johanstokking** [comment]
> > Hi Jaime,
> > Until 4th Batch Update, there are only 2 models left from us due to unsure information, they are UC50x and UC51x. Besides there are discontinued products UC1114, UC1122, UC1152, there might be customers using them on TTN but we don't sell them or have their product URL any more. Would you suggest that I should create the files for them and leave URLs blank? Waiting for your reply, thanks.
> 
> Hi @johanstokking, issues about the filename should have been resolved. Could you please check out this question about Milesight discontinued products?

Thanks. Yes, please make sure that you don't remove products from the Device Repository. Indeed, if you don't sell them anymore, you can remove URLs. We may add a field to indicate that the device has been discontinued or superseded by another device in the future.

This branch is still not rebased. Please rebase on latest master.

**Comment by Milesight-IoT** [comment]
Hi @johanstokking I've rebased on latest master, will add the discontinued devices soon.

**Comment by johanstokking** [comment]
Thanks @Milesight-IoT. @Jaime-Trinidad will pick this up when you're done.
